digdag-core/src/main/java/io/digdag/core/database/DatabaseSessionStoreManager.java

```diff
@@ -871,20 +872,31 @@ public boolean copyInitialTasksForRetry(List<Long> recursiveChildrenIdList)
         {
             List<StoredTask> tasks = handle.createQuery(
                     selectTaskDetailsQuery() + " where t.id " + inLargeIdListExpression(recursiveChildrenIdList) +
-                    " and " + bitAnd("t.state_flags", Integer.toString(TaskStateFlags.INITIAL_TASK)) + " != 0"  // only initial tasks
+                    " and " + bitAnd("t.state_flags", Integer.toString(TaskStateFlags.INITIAL_TASK)) + " != 0" + // only initial tasks
+                    " order by t.id asc" // to ensure that tasks are sorted in the newest order
                 )
                 .map(stm)
                 .list();
             if (tasks.isEmpty()) {
                 return false;
             }
+
+            DatabaseTaskControlStore store = new DatabaseTaskControlStore(handle);
+            Map<Long, Long> oldIdToNewId = new HashMap<>();
             for (StoredTask task : tasks) {
                 Task newTask = Task.taskBuilder()
                     .from(task)
                     .state(TaskStateCode.BLOCKED)
                     .stateFlags(TaskStateFlags.empty())
                     .build();
-                addSubtask(tasks.get(0).getAttemptId(), newTask);
+                long newTaskId = addSubtask(tasks.get(0).getAttemptId(), newTask);
+
+                // addSubtask doesn't copy fields of StoredTask because its second argument is Task.
+                // Copy StoredTask::getUpstreams here
+                oldIdToNewId.put(task.getId(), newTaskId);
+                for (long oldUpstreamId : task.getUpstreams()) {
```

Comment:
1. It seems the task ID mapping table `oldIdToNewId` isn't completely built up here. So new dependencies can use old task ID, I guess. Don't we need to build `oldIdToNewId` first and then add dependencies?

2. `DatabaseTaskControlStore#addDependencies` is a more abstract API, so we'd better use it?

IIUC, those codes can be like this.
```java
            Map<Long, Long> oldIdToNewId = new HashMap<>();
            for (StoredTask task : tasks) {
                Task newTask = Task.taskBuilder()
                        .from(task)
                        .state(TaskStateCode.BLOCKED)
                        .stateFlags(TaskStateFlags.empty())
                        .build();
                long newTaskId = addSubtask(tasks.get(0).getAttemptId(), newTask);

                // Build task id mapping table
                oldIdToNewId.put(task.getId(), newTaskId);
            }

            for (StoredTask task : tasks) {
                long newTaskId = oldIdToNewId.get(task.getId());

                // addSubtask doesn't copy fields of StoredTask because its second argument is Task.
                // Copy StoredTask::getUpstreams here
                store.addDependencies(newTaskId,
                        task.getUpstreams().stream()
                        .map(
                                upstreamId -> {
                                    Long newUpstreamId = oldIdToNewId.get(upstreamId);
                                    if (newUpstreamId == null) {
                                        return upstreamId;
                                    }
                                    else {
                                        return newUpstreamId;
                                    }
                                })
                        .collect(Collectors.toList()));
            }
```

Reply:
1.

Dependency is possible only from a later task to a former task. Dependent task is is always smaller. So we don’t have to prebuild the mapping.

2.

I also thought it first but I thought that dao.insertTaskDependency is already well abstracted